third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -125,106 +168,209 @@ impl TransactionDataCache {
+            let (data, bytes_loaded) = {
+                // If we need to process delayed fields, we pass type layout to remote storage. Remote
+                // storage, in turn ensures that all delayed field values are pre-processed.
+                resource_resolver.get_resource_bytes_with_metadata_and_layout(
+                    addr,
+                    struct_tag,
+                    &metadata,
+                    if contains_delayed_fields {
+                        Some(layout)
+                    } else {
+                        None
+                    },
+                )?
+            };
+            let function_value_extension = FunctionValueExtensionAdapter { module_storage };
+            let value = match data {
+                Some(blob) => {
+                    let max_value_nest_depth = function_value_extension.max_value_nest_depth();
+                    let val = ValueSerDeContext::new(max_value_nest_depth)
+                        .with_func_args_deserialization(&function_value_extension)
+                        .with_delayed_fields_serde()
+                        .deserialize(&blob, layout)
+                        .ok_or_else(|| {
+                            let msg = format!(
+                                "Failed to deserialize resource {} at {}!",
+                                struct_tag.to_canonical_string(),
+                                addr
+                            );
+                            PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_RESOURCE)
+                                .with_message(msg)
+                        })?;
+                    GlobalValue::cached(val)?
+                },
+                None => GlobalValue::none(),
+            };
+            (CachedInformation::Value(value), bytes_loaded)
+        } else {
+            let (size, bytes_loaded) = resource_resolver
+                .get_resource_size_with_metadata_and_layout(
+                    addr,
+                    struct_tag,
+                    &metadata,
+                    if contains_delayed_fields {
+                        Some(layout)
+                    } else {
+                        None
+                    },
+                )?;
+            (CachedInformation::SizeOnly(size), bytes_loaded)
+        })
+    }
+
     /// Retrieves data from the remote on-chain storage and converts it into a [DataCacheEntry].
     /// Also returns the size of the loaded resource in bytes. This method does not add the entry
     /// to the cache - it is the caller's responsibility to add it there.
-    pub(crate) fn create_data_cache_entry(
+    /// If `load_data` is false, only resource existence information will be retrieved
+    pub(crate) fn create_and_insert_or_upgrade_and_charge_data_cache_entry(
+        &mut self,
         module_storage: &dyn ModuleStorage,
         resource_resolver: &dyn ResourceResolver,
+        maybe_gas_meter: Option<&mut impl GasMeter>,
```

Comment:
I remember @georgemitenkov had a kind-of no-op gas meter too - maybe a little overhead of passing method call / params. I am okay w. option though. But maybe if we refactor it will not be a question anyway.